aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -1018,6 +1018,23 @@ Tensor& heaviside_(Tensor& self, const Tensor& values) {
   return at::heaviside_out(self, self, values);
 }
 
+Tensor& ldexp_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(!self.is_complex(), "ldexp does not support complex tensors.");
+  TORCH_CHECK(result.dtype() == self.dtype(), "ldexp result dtype must match input dtype.");
+  native::mul_out(result, self, at::pow(2.0, other));
+  return result;
+}
+
+Tensor ldexp(const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(!self.is_complex(), "ldexp does not support complex tensors.");
+  Tensor result = at::empty({0}, self.options());
```

Comment:
`Tensor result;`

If you create the tensor like this it will have self's dtype, but now that the op is implemented as a composite it inherits mul's type promotion, so the result dtype may not be self's.

Reply:
If I don't initialize result at all i get an error like this:
`RuntimeError: Expected a Tensor of type Variable but found an undefined Tensor for argument #0 'out'` if using `at::mul_out`

Any idea how to properly create result?

Reply:
Oh right, sorry, I always forget we have this wonky issue with TensorIterator. This is also why mul and mul_out have divergent implementations:

https://github.com/pytorch/pytorch/blob/ff0af7242b6842f4c05bdd8737ccc40791510e91/aten/src/ATen/native/BinaryOps.cpp#L243

My mistake on that. You can just write this as return self * 2 ** other here, too. Also I think the checks for complex can be dropped now, too, since we don't support complex multiplication.

Reply:
I tried using just `self * at::pow(2.0, other)` and hit some type casting errors - want to take a look at what I ended up with and let me know if we're good to go?


```c++
Tensor& ldexp_out(Tensor& result, const Tensor& self, const Tensor& other) {
  return at::mul_out(result, self, at::pow(2.0, other));
}

Tensor ldexp(const Tensor& self, const Tensor& other) {
  return at::mul(self, at::pow(2.0, other));
}

Tensor& ldexp_(Tensor& self, const Tensor& other) {
  return at::ldexp_out(self, self, other);
}
```

Reply:
This looks correct to me -- what were the type casting errors?

Reply:
Oh it was the implementation with literal `*`  instead `mul`

Reply:
Oh, right. Got it. Thanks for explaining that to me again ;)